liblastlog2/src/lastlog2.c

```diff
@@ -0,0 +1,574 @@
+int ll2_check_database (const char *lastlog2_path)
+{
+	struct stat st;
+	return stat(lastlog2_path, &st);
+}
+
+/* Reads one entry from database and returns that.
+   Returns 0 on success, -1 on failure. */
+static int
+read_entry (sqlite3 *db, const char *user,
+	    int64_t *ll_time, char **tty, char **rhost,
+	    char **pam_service, char **error)
+{
+	int retval = 0;
+	sqlite3_stmt *res;
+	char *sql = "SELECT * FROM Lastlog2 WHERE Name = ?";
+
+	if (sqlite3_prepare_v2 (db, sql, -1, &res, 0) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to execute statement: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup ("Out of memory");
+
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 1, user, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create search query: %s",
+				      sqlite3_errmsg (db)) < 0)
+				*error = strdup("Out of memory");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	int step = sqlite3_step (res);
+
+	if (step == SQLITE_ROW) {
+		const unsigned char *luser = sqlite3_column_text (res, 0);
+		const unsigned char *uc;
+
+		if (strcmp ((const char *)luser, user) != 0) {
+			if (error)
+				if (asprintf (error, "Returned data is for %s, not %s", luser, user) < 0)
+					*error = strdup("Out of memory");
+
+			sqlite3_finalize (res);
+			return -1;
+		}
+
+		if (ll_time)
+			*ll_time = sqlite3_column_int64 (res, 1);
+
+		if (tty) {
+			uc = sqlite3_column_text (res, 2);
+			if (uc != NULL && strlen ((const char *)uc) > 0)
+				*tty = strdup ((const char *)uc);
+		}
+		if (rhost) {
+			uc = sqlite3_column_text (res, 3);
+			if (uc != NULL && strlen ((const char *)uc) > 0)
+				*rhost = strdup ((const char *)uc);
+		}
+		if (pam_service) {
+			uc = sqlite3_column_text (res, 4);
+			if (uc != NULL && strlen ((const char *)uc) > 0)
+				*pam_service = strdup ((const char *)uc);
+		}
+	} else {
+		if (error)
+			if (asprintf (error, "User '%s' not found (%d)", user, step) < 0)
+				*error = strdup("Out of memory");
+
+		retval = -1;
+	}
+
+	sqlite3_finalize (res);
+
+	return retval;
+}
+
+/* reads 1 entry from database and returns that. Returns 0 on success, -1 on failure. */
+int
+ll2_read_entry (const char *lastlog2_path, const char *user,
+		int64_t *ll_time, char **tty, char **rhost,
+		char **pam_service, char **error)
+{
+	sqlite3 *db;
+	int retval;
+
+	if ((db = open_database_ro (lastlog2_path, error)) == NULL)
+		return -1;
+
+	retval = read_entry (db, user, ll_time, tty, rhost, pam_service, error);
+
+	sqlite3_close (db);
+
+	return retval;
+}
+
+/* Write a new entry. Returns 0 on success, -1 on failure. */
+static int
+write_entry (sqlite3 *db, const char *user,
+	     int64_t ll_time, const char *tty, const char *rhost,
+	     const char *pam_service, char **error)
+{
+	char *err_msg = NULL;
+	sqlite3_stmt *res;
+	char *sql_table = "CREATE TABLE IF NOT EXISTS Lastlog2(Name TEXT PRIMARY KEY, Time INTEGER, TTY TEXT, RemoteHost TEXT, Service TEXT) STRICT;";
```

Comment:
The version of libsqlite used by Ubuntu 18.04 in the test suites fails with:
```
SQL error: near "STRICT": syntax error
```

Reply:
I have removed STRICT.